### PR TITLE
Enhance DecomposeUnsupportedConversions.

### PR DESCRIPTION
**Pull request by chengjunlu**
1. Handle the Intel DPAS layout.
2. Fix issue in block -> dot layout decompose.

### CODE REVIEW

File: test/TritonIntelGPU/tritonintlgpu-decompose-unsupported-cvt-layout.mlir [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+    tt.return
+  }
+}
+
+// -----
+
+#blocked = #triton_gpu.blocked<{sizePerThread = [4, 1], threadsPerWarp = [16, 1], warpsPerCTA = [1, 32], order = [0, 1]}>
+#dot_a = #triton_gpu.dot_op<{opIdx = 0, parent = #blocked}>
+#dot_b = #triton_gpu.dot_op<{opIdx = 1, parent = #blocked}>
+module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32 : i32, triton_gpu.target = "xpu:DEVICE_ARCH.PVC", "triton_gpu.threads-per-warp" = 16 : i32, triton_intel_gpu.min_sg_size = 16 : i32, triton_intel_gpu.support_dpas, triton_intel_gpu.support_sg_2d_block} {
```
Comment (whitneywhtsang) [inline test/TritonIntelGPU/tritonintlgpu-decompose-unsupported-cvt-layout.mlir:25]:
```suggestion
module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32 : i32, "triton_gpu.threads-per-warp" = 16 : i32, triton_intel_gpu.min_sg_size = 16 : i32, triton_intel_gpu.support_dpas, triton_intel_gpu.support_sg_2d_block} {
```

File: test/TritonIntelGPU/tritonintlgpu-decompose-unsupported-cvt-layout.mlir [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+#dot_b = #triton_gpu.dot_op<{opIdx = 1, parent = #dpas, kWidth = 2}>
+module attributes {"triton_gpu.num-ctas" = 1 : i32, "triton_gpu.num-warps" = 32 : i32, "triton_gpu.threads-per-warp" = 16 : i32, triton_intel_gpu.min_sg_size = 16 : i32, triton_intel_gpu.support_dpas, triton_intel_gpu.support_sg_2d_block} {
+  // CHECK-LABEL: decompose_dpas_2_dot
+  tt.func public @decompose_dpas_2_dot() {
+    %cst = arith.constant dense<0.000000e+00> : tensor<64x32xf16, #dpas>
+    // CHECK-NOT: triton_gpu.convert_layout
+    // CHECK: triton_gpu.local_alloc
+    // CHECK: triton_gpu.local_load
+    %0 = triton_gpu.convert_layout %cst : tensor<64x32xf16, #dpas> -> tensor<64x32xf16, #dot_b>
+    // CHECK: triton_gpu.convert_layout
```
Comment (whitneywhtsang) [inline test/TritonIntelGPU/tritonintlgpu-decompose-unsupported-cvt-layout.mlir:14]: Please add comment in the test case on why you don't expect local_alloc/load in this case.

File: third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp [hunk 3]
```diff
@@ -133,12 +138,10 @@ struct DecomposeUnsupportedConversions
...
             sharedMemorySpace);
-        auto tmp = builder.create<triton::gpu::ConvertLayoutOp>(
+        auto tmp = builder.create<triton::gpu::LocalAllocOp>(
             cvtOp.getLoc(), tmpType, cvtOp.getSrc());
-        addAttrs(tmp, cvtOp->getAttrs());
-        auto newConvert = builder.create<triton::gpu::ConvertLayoutOp>(
+        auto newConvert = builder.create<triton::gpu::LocalLoadOp>(
             cvtOp.getLoc(), dstType, tmp);
-        addAttrs(newConvert, cvtOp->getAttrs());
         cvtOp.replaceAllUsesWith(newConvert.getResult());
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp:145]: Please upstream this changes.

File: test/TritonIntelGPU/tritonintlgpu-decompose-unsupported-cvt-layout.mlir (hunk 2 above)
Comment (chengjunlu) [reply test/TritonIntelGPU/tritonintlgpu-decompose-unsupported-cvt-layout.mlir:16]: Added comments.

**Review by whitneywhtsang — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp [hunk 4]
```diff
@@ -84,29 +86,32 @@ struct DecomposeUnsupportedConversions
       }
     });
     /* -------------------------------- */
-    // Replace `mma -> dot_op` with `mma -> blocked -> dot_op`
+    // Replace `dpas -> dot_op` with `dpas -> shared -> dot_op`
```
Comment (etiotto) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp:89]: Why not doing the same as NVidia and go to blocked layout which is later on converted to shared layout  and then to dot layout. I think your motivation is to just avoid the intermediate step of converting from a dpas layout to blocked layout, which makes sense at first sight. 

My concern is that deviating from the way NVidia is doing layout conversion might expose problems in other places. So I'd like to understand your rationale for this "shortcut" better.

Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp:89]: The reason is for the performance concern. There is extra conversion thru `dpas->blocked->shared->dot_op`. 
And this pass is the last pass before lowering to LLVM dialect. No further pass to fold the conversion.

But the most distributed->dot_op conversion has been rewritten in the `ReduceDataDuplication` pass. I am Ok to keep the NV way.

**Review by etiotto — APPROVED** [review]

File: third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp [hunk 5]
```diff
@@ -84,23 +86,24 @@ struct DecomposeUnsupportedConversions
...
```
Comment (whitneywhtsang) [inline third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp:89]:
```suggestion
    // Replace `dpas -> dot_op` with `dpas -> blocked -> dot_op`
```

File: third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp (hunk 3 above)
Comment (whitneywhtsang) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp:143]: @chengjunlu Please create issue to track upstreaming this change.

Comment (chengjunlu) [reply third_party/intel/lib/TritonIntelGPUToLLVM/DecomposeUnsupportedConversions.cpp:143]: https://github.com/intel/intel-xpu-backend-for-triton/issues/1796

**Pull request merged**